### PR TITLE
Handle refreshing the Auth0 token

### LINKED ISSUE #155: Handle refreshing the Auth0 token
The Auth0 token is set to expire within 2 hours and we will need to figure out how to refresh the token.

### PR DESCRIPTION
**Pull request by peterbe**
fixes #155

### CODE REVIEW

**Comment by peterbe** [comment]
See the changes I made to the `dev.rst` about how to test this.

File: docs/config.rst [hunk 1]
```diff
@@ -41,3 +41,19 @@ An environment variable like ``REACT_APP_FOO`` controls the setting ``FOO``.
...
+    :default: 300
+
+    How often to issue a silent authentication refresh.
+
+.. envvar:: REACT_APP_REFRESH_PREEMPTIVE_SECONDS
+
+    :default: 900
+
+    How soon *before* the session expires do we perform a silent authentication to keep it fresh.
+    Setting this liberally means we can make sure the access token is up-to-date *before* the
```
Comment (magopian) [inline docs/config.rst]: is this a typo, and should it read `literally` instead of `liberally`? Not sure...

File: docs/config.rst [hunk 2]
```diff
@@ -41,3 +41,19 @@ An environment variable like ``REACT_APP_FOO`` controls the setting ``FOO``.
...
+    user attempts to submit something to a backend and then avoid the access token being
+    out of date.
```
Comment (magopian) [inline docs/config.rst]: I don't understand how both settings interact with each other. Are you saying we need a new silent auth refresh every 300 seconds, but also 900 seconds before the session expires?
It seems weird to have the first one being shorter than the second one, which feels like the second one is useless?

Comment (magopian) [reply docs/config.rst]: Or maybe it's a naming issue? Is it "how often should we check if it's time to refresh" for the first var, and "how long before the expiry is it time to refresh"?

So maybe they could be named something like `REACT_APP_CHECK_SESSION_EVERY_SECONDS` and `REACT_APP_REFRESH_BEFORE_SECONDS`?

I still feel those two could be one and unique, I'm not sure I see the benefit in having both.

File: docs/config.rst (hunk 1 above)
Comment (peterbe) [reply docs/config.rst]: I meant liberally. But now I see that that's a poor choice of words. The intention was to say "settings this QUITE HIGH means..." or something like that. I'll think of a better way of phrasing it.

File: docs/config.rst (hunk 2 above)
Comment (peterbe) [reply docs/config.rst]: Yeah, the intention is that there are two numbers. One is how often long the sleep is in the infinite loop, the other is what is too close to the end of the expires. Inside the code I use the word "preemptive" which is basically a way to make sure the accesstoken is refreshed *before* it expires. To your point, the word "before" is probably better. 

I'll try to review the language and update.

Comment (peterbe) [reply docs/config.rst]: @magopian Have a look again. I only changed the documentation text. Is it clearer now?

File: docs/config.rst [hunk 3]
```diff
@@ -41,3 +41,20 @@ An environment variable like ``REACT_APP_FOO`` controls the setting ``FOO``.
...
+
+.. envvar:: REACT_APP_REFRESH_AUTH_SESSION_SECONDS
+
+    :default: 300
+
+    How often to issue a silent authentication refresh. Technically, when you're logged in,
+    an infinite loop is run that refreshes your access token forever. And this number is
+    number of seconds to sleep between each check.
+
+.. envvar:: REACT_APP_REFRESH_PREEMPTIVE_SECONDS
```
Comment (mythmon) [inline docs/config.rst]: This variable name doesn't mention auth or sessions. Perhaps `REACT_APP_AUTH_REFRESH_PREEMPTIVE_SECONDS`? This is all getting really long though. Mostly I think it should mention auth in some way.

PS: I really like that y'all are putting units in the setting names. :+1:

File: docs/config.rst [hunk 4]
```diff
@@ -41,3 +41,20 @@ An environment variable like ``REACT_APP_FOO`` controls the setting ``FOO``.
     :default: ``https://${OIDC_DOMAIN}/userinfo``
 
    The audience for the access token that is generated by Auth0.
+
+
+.. envvar:: REACT_APP_REFRESH_AUTH_SESSION_SECONDS
```
Comment (mythmon) [inline docs/config.rst]: Perhaps specifically mentioning this is the period of the timer would help? Maybe something like `REACT_APP_AUTH_REFRESH_PERIOD_SECONDS`?

File: docs/config.rst (hunk 3 above)
Comment (peterbe) [reply docs/config.rst]: Good point @mythmon about the need for mentioning "AUTH" in this one.

I'd rather it be long  than unclear.

@magopian Can you second mythmon's point and I'll make a final pass on this.

**Comment by peterbe** [comment]
Ok. Updated as per mixed feedback. Now both new env vars correctly have the "_AUTH_" part in the name. And, for the setTimeout loop, instead of "_SESSION_" it's called "_PERIOD_".

File: docs/config.rst [hunk 5]
```diff
@@ -41,3 +41,20 @@ An environment variable like ``REACT_APP_FOO`` controls the setting ``FOO``.
     :default: ``https://${OIDC_DOMAIN}/userinfo``
 
    The audience for the access token that is generated by Auth0.
+
+
+.. envvar:: REACT_APP_REFRESH_AUTH_PERIOD_SECONDS
+
+    :default: 300
+
+    How often to issue a silent authentication refresh. Technically, when you're logged in,
```
Comment (magopian) [inline docs/config.rst]: It's not how often to issue a silent authentication refresh, but how often do we check if the expiry is in less than `REACT_APP_REFRESH_AUTH_PREEMPTIVE_SECONDS`.

File: docs/config.rst [hunk 6]
```diff
@@ -41,3 +41,20 @@ An environment variable like ``REACT_APP_FOO`` controls the setting ``FOO``.
...
+    an infinite loop is run that refreshes your access token forever. And this number is
+    number of seconds to sleep between each check.
+
+.. envvar:: REACT_APP_REFRESH_AUTH_PREEMPTIVE_SECONDS
+
+    :default: 900
+
+    How soon *before* the session expires do we perform a silent authentication to keep it fresh.
+    If you set this really low (or even to zero) the risk is that the access token has expired
+    by the time it's needed for an API endpoint call.
```
Comment (magopian) [inline docs/config.rst]: I would reword this last sentence along the lines of
```
If you set this lower than `REACT_APP_REFRESH_AUTH_PERIOD_SECONDS` then the access token might have expired before it was detected that it needed a refresh.
```

**Comment by peterbe** [comment]
How about, instead of  `REACT_APP_REFRESH_AUTH_PERIOD_SECONDS` we call it `REACT_APP_REFRESH_AUTH_CHECK_SECONDS`? "Period" is refers to the length and "check" refers to the action.

File: docs/config.rst (hunk 6 above)
Comment (peterbe) [reply docs/config.rst]: I looked at it and I think the previous sentence was easier to read. 
Your suggestion makes it feel like there are only two single things happening. The check if the access token has expired and the number of seconds before the access token expires. 

What it actually is is that the REACT_APP_REFRESH_AUTH_PERIOD_SECONDS is used all the time, repeatedly. 99% of the time nothing happens. But 1% of the time the business logic deems the access token expired (or about to expire) and thus needs to be refreshed.

**Comment by mythmon** [comment]
> How about having only one variable (how often to check for the need to refresh), and then in the code, check if the access token is expiring in less than 2* to avoid not detecting the expiry soon enough (or 1.5, or 3 if you want to be conservative, whatever, or <2 minutes> + ).

I like this idea. There is a lot of interaction between these two settings, and in practice they should probably be related. For example, if the preemptive one is shorter than the check setting, that's going to be bad. Removing the preemptive setting, and automatically deriving it from the check period timer makes a lot of sense to me.

**Comment by peterbe** [comment]
I kinda wish I had done it like this:

```javascript
recheckAccessToken() {

  // come back again soon no matter how the rest goes!
  window.setTimeout(this.recheckAccessToken, 30 * 1000) 

  const expires = int(localStorage.expiresAt)
  if (expires - 60 * 1000 - new Date().getTime() < 0) {
    webAuth.checkSession(...)
  }
}
```

I.e. the number of seconds to loop and the number of seconds to look ahead all hardcoded straight in the JS code. But instead, I moved these numbers out into `settings.js` and tried to give them names. That clearly failed :)

Having them as constants and even having them derived from possible env vars makes things more flexible but it also adds more noise and confusion.

**Comment by peterbe** [comment]
ping @rehandalal for second opinion/review. I think the branch works really well. It'just that I can't seem to nail the name of the constants or the documentation verbiage.

**Comment by peterbe** [comment]
Actually, I've found a bug. Maybe. See [this comment](https://github.com/mozilla/delivery-console/issues/248#issuecomment-400656725) which concludes that the silent authentication can "revive" a dead access token. That's great, but there's a subtle flaw in the whole auth + refresh logic. 

If you open a page, first thing in the morning, like `/recipe/123/edit` then two `componentDidMount` are going to run. One for the auth stuff and one for the Recipe stuff (technically `QueryRecipe` which calls `fetchRecipe`). So this **timeline** might happen:

1. `fetch('/normandy/api/recipes/123', {Authorization: 'Bearer abc123'})`   # Promise 1
2. `webAuth.checkSession()` (because it believes access token `abc123` is expired)  # Promise 2
3. Promise 1 resolves (with a 401 Unauthorized from Normandy's API)
4. Promise 2 resolves (and it updates the localStorage with a fresh new accessToken)

I suspect that the best thing we can do is to fix the whole auth flow but with some tricks. Right here: https://github.com/peterbe/delivery-console/blob/328cfd2aabaa538c7977a9058f217f79ce76abf3/src/components/data/QueryAuth0.js#L45
we're making the mistake of saying "Oh, there's already an access token in localStorage." What we should do is saying "Oh, there's already an access token in localStorage, but, is it expired? If it's expired don't bother."

Clear as mud? 

I'm going to try to reproduce it. The reason I'm hesitant is that I'm not 100% clear on how the access token travels from `localStorage` to Redux to `QueryRecipe.js`.

**Comment by peterbe** [comment]
@rehandalal Fixed up from master. Works locally.

File: src/components/data/QueryAuth0.js [hunk 7]
```diff
@@ -37,24 +38,82 @@ export default class QueryAuth0 extends React.PureComponent {
...
+    const preemptive = REFRESH_AUTH_PREEMPTIVE_SECONDS * 1000;
 
-      if (state) {
-        this.props.push(state);
+    const accessTokenRefreshLoopTimer = window.setTimeout(async () => {
+      await this.accessTokenRefreshLoop();
+    }, REFRESH_AUTH_PERIOD_SECONDS * 1000);
+    if (left - preemptive < 0) {
+      // Time to refresh!
+      console.warn('Time to refresh auth session');
```
Comment (rehandalal) [inline src/components/data/QueryAuth0.js]: I don't think this should be a warning maybe:
```js
console.log('Refreshing the Auth0 access token');
```

File: src/components/data/QueryAuth0.js [hunk 8]
```diff
@@ -37,24 +38,82 @@ export default class QueryAuth0 extends React.PureComponent {
...
+      try {
+        const authResult = await refreshAuthentication(window.location.pathname);
+        this.postProcessAuthResult(authResult);
+      } catch (err) {
+        window.clearTimeout(accessTokenRefreshLoopTimer);
+        console.error(err);
```
Comment (rehandalal) [inline src/components/data/QueryAuth0.js]: I don't think we need to put this in here since we're handling this error on the next line.

File: src/components/data/QueryAuth0.js [hunk 9]
```diff
@@ -37,24 +38,82 @@ export default class QueryAuth0 extends React.PureComponent {
...
+    // included with the accessToken. Use this now to update the state so that we
+    // can display the name and avatar you logged in as.
+    userProfileReceived(authResult.idTokenPayload);
+
+    if (state) {
+      this.props.push(state);
+    }
+  };
+
+  async accessTokenRefreshLoop() {
```
Comment (rehandalal) [inline src/components/data/QueryAuth0.js]: I think we can ditch the second setting if we refactor this slightly:
```js
async accessTokenRefreshLoop() {
  if (!localStorage.getItem('expiresAt')) {
    return;
  }
  const expiresAt = JSON.parse(localStorage.getItem('expiresAt'));
  const expiresIn = expiresAt - new Date().getTime();
  const shouldRefresh = expiresIn <= REFRESH_AUTH_PERIOD_SECONDS;

  const accessTokenRefreshLoopTimer = window.setTimeout(async () => {
    await this.accessTokenRefreshLoop();
  }, REFRESH_AUTH_PERIOD_SECONDS * 1000);

  if (shouldRefresh) {
```

File: src/components/data/QueryAuth0.js [hunk 10]
```diff
@@ -37,24 +38,82 @@ export default class QueryAuth0 extends React.PureComponent {
...
+    // By default this is true if the window.location.hash parsing stuff worked.
+    // We might change our minds about this later.
+    let startAccessTokenRefreshLoop = !!authResult;
+
     if (!authResult) {
       // It if was null from finishAuthenticationFlow() it means it didn't pick anything
       // up from the window.location.hash so we can potentially use what was stored in
       // localStorage.
       authResult = JSON.parse(localStorage.getItem('authResult'));
+      if (authResult) {
```
Comment (rehandalal) [inline src/components/data/QueryAuth0.js]: this could just be:
```js
startAcessTokenRefreshLoop = !!authResult;
```

File: src/components/data/QueryAuth0.js [hunk 11]
```diff
@@ -37,24 +38,82 @@ export default class QueryAuth0 extends React.PureComponent {
...
+        // If you have previously logged in we'll want to start the accessToken refresh
+        // loop. Even if it has expired. Because, if it indeed as expired, the
+        // accessToken refresh loop can silently authenticate you again.
+        startAccessTokenRefreshLoop = true;
+      }
+      // But this is only good enough if it hasn't expired.
+      const expiresAt = JSON.parse(localStorage.getItem('expiresAt'));
+      if (expiresAt) {
```
Comment (rehandalal) [inline src/components/data/QueryAuth0.js]: This should be collapsed to:
```js
if (expiresAt && expiresAt - new Date().getTime() < 0) {
```

File: src/settings.js [hunk 12]
```diff
@@ -12,3 +12,11 @@ export const OIDC_DOMAIN = process.env.REACT_APP_OIDC_DOMAIN || 'auth.mozilla.au
 export const OIDC_CALLBACK_URL = process.env.REACT_APP_OIDC_CALLBACK_URL || window.location.origin;
 export const OIDC_AUDIENCE =
   process.env.REACT_APP_OIDC_AUDIENCE || `https://${OIDC_DOMAIN}/userinfo`;
+export const REFRESH_AUTH_PERIOD_SECONDS = parseInt(
```
[the remark this answers is not in the document]
Comment (peterbe) [reply src/settings.js]: I'll take it!

File: src/components/data/QueryAuth0.js (hunk 10 above)
Comment (peterbe) [reply src/components/data/QueryAuth0.js]: I don't see that. We need to evaluate the value of `authResult` **twice**. 
If it was something true'y after the `parseHash` we should start the refresh loop. But also, if it was falsy after `parseHash` and we *did* have something in `localStorage` that should also start the refresh loop. 

Can you take a second look? Perhaps my code comments weren't clear.

Comment (rehandalal) [reply src/components/data/QueryAuth0.js]: ah! my bad. I didn't think about the fact that this was already assigned a value above.

this could be:
```js
startAccessTokenRefreshLoop = authResult ? true : startAccessTokenRefreshLoop;
```
